core/src/main/java/org/apache/iceberg/actions/RewriteFileGroupPlanner.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.actions;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Stream;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.RewriteJobOrder;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.data.GenericRecord;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.StructLikeMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Checks the files in the table, and using the {@link FileRewriter} plans the groups for
```

Comment:
I'm a little confused about the inheritance structure here. If we want to make this common, maybe we should just have this belong to the FileRewriter as a base class? 

If I can think through this correctly in this draft we have

RewriteFileGroupPlanner Responsible for 
1. Scanning Table
2. Grouping Tasks by Partition
3. Further Grouping and Filtering in FileRewriter

FileRewriter Responsible for
1. Filtering Tasks based on Rewriter Config
2. Grouping tasks within partitions

Feels like all of that could just be in the FileRewriter without having a separate GroupPlanner? Do we have an argument against that?

Reply:
Let me check that. I was hesitant to change existing classes, but if we have a consensus, I'm happy to do that.

Reply:
The `FileRewriter` is a generic class with `<T extends ContentScanTask<F>, F extends ContentFile<F>>`.
The `RewriteFileGroup` which is returned by the planning is specific for `FileScanTask`.

I'm not sure that we use the `FileRewriter` with any other parametrization ATM, but to put everything into a single class we need to do one of the things below:
1. Changing `FileRewriter<T extends FileScanTask>` instead of the current `FileRewriter<T extends ContentScanTask<F>, F extends ContentFile<F>>`, or 
2. Changing the `public List<FileScanTask> RewriteFileGroup.fileScans()` to return `public <T extends ContentScanTask<F>, F extends ContentFile<F>>  List<T> fileScans`

Both of the would be a breaking change in the public API so we should be careful about them.

We still can "duplicate" and create a new FileRewriter, and deprecate the old one - but I would only do this, if we think that the original API needs a serious refactor anyways.

WDYT?

Reply:
I've been thinking a lot about the refactor. Trying to figure out what we should do.

we could possibly do

Rewriter[T]
  has a Planner[T] // Responsible for generating tasks (move actual filtering into rewriter)
Maybe we should also have
```
Planner[T]
    DeletePlanner[DeleteScanTask]
    FilePlanner[FileScanTask]
```    
    
So then you would have
```
SizeBasedDataRewriter <FileScanTask, DataFile>
   private planner = new FilePlanner<FileScanTask>
```

Would that end up being more complicated? I feel like we are getting some interdependencies here I'm not comfortable with extending. @aokolnychyi Probably has some smart things to say here as well